### ISSUE TITLE
Failing to load collision integrals when condensed species are present

### ISSUE THREAD

**Issue by turchia** [mutationpp/Mutationpp#34]
### Description
mppequil seems to fail when a condensed species is present in the mixture because it's trying to load collision integral data for pairs containing the condensed species (e.g., Q11_(C,C(gr)))

### Error text
  mppequil -T 300 -P 10135 -m 0,32  carbonPhenolInAir
         Th[K]      mu[Pa-s]
           200
M++ error: error parsing file.
file: /personnel/reseng_ar/turchi/PROGRAMS/github_mpp/Mutationpp/data/transport/collisions.xml
line: 52
Could not find data necessary to load Q11_(C,C(gr)).
Was trying to load a collision integral.
[carbonPhenolInAir.txt](https://github.com/mutationpp/Mutationpp/files/2844845/carbonPhenolInAir.txt)

P.S.
This thing doesn't like .xml files. You should change all the M++ structure to avoid .xml files

**Comment by jbscoggi** [comment]
Hi @turchia, can you explain what the use-case is here?  I'm not sure it makes sense to compute transport of a multiphase mixture...  Is the intended behavior to compute transport of the gas phase only?

Not sure why github doesn't support attaching XML files, but if you want in the future you can add xml directly in the issue:

```xml
<mixture thermo_db="NASA-9" state_model="EquilTP" mechanism="none">
  <default_element_fractions>
    C : 0.125,
    H : 0.750,
    O : 0.125,
    N : 0
  </default_element_fractions>

  <species>
    C(gr) C H O N CH CH4 CO CO2 CN C2H C2H2,acetylene C3 C3H C4 
    C4H2,butadiyne C5 HCN H2 H2O N2
  </species>

  <element_compositions>
    <composition name="VKIZuramPyroGas">
      C:0.125, H:0.750, O:0.125, N:0.0
    </composition>
    <composition name="BLedge">
      C:0., H:0., O:0.21, N:0.79
    </composition>
    <composition name="Char">
      C:1., H:0., O:0., N:0.0
    </composition>
  </element_compositions>
</mixture>
```

**Comment by turchia** [comment]
Hi @jbscoggi,

You got the point right

> I'm not sure it makes sense to compute transport of a multiphase mixture...

Fully agree

>  Is the intended behavior to compute transport of the gas phase only?

Yes, that's exactly the intention

But the file contains the solid phase because it will be used in a material code and the C(gr) "could be useful" for the B' computation.

**Comment by jbscoggi** [comment]
Hey @turchia, I took a look at the transport database stuff and I think I have a "simple" fix for this problem.  Obviously, what we want is to loop over only gas phase species when we compute transport properties.  Fortunately, the ordering of species in Thermodynamics is fixed to ensure that gaseous species are always first, and there is a function `nGas()` which provides the number of gas species in the mixture.  In addition, the transport classes rely on the `CollisionDB::nSpecies()` function to provide the number of species in the mixture, which currently copies `Thermodynamics::nSpecies()`.  It should be easy enough to change that behavior so that `CollisionDB::nSpecies()` returns `Thermodynamics::nGas()` instead.  Unfortunately, this isn't the whole story because it seems that in some places in transport, we are still calling the`nSpecies()` directly from the Thermodynamics, which means we need to make sure that we change all of these occurrences accordingly, but I don't think that should be too much work.

So, the next step is to figure out who will do this :).  Like I said, it is not a big work, but perhaps this is a good training opportunity for someone to go through the process of making some changes and merging it back to the code base.  I can also do it myself, but you will have to wait a few days because I have several outstanding things to work on.

I think this problem in general is also a good motivation for restructuring the high level classes in a future release, to make it more natural how we handle multiple phases.  For example, in Cantera, phases are objects which each have their own thermo, transport, etc.  This might be a good model to consider in the future.

Lastly, I realized that the mixture file you uploaded still has the old default element format.  I have cleaned it up below with the new format for specifying the default:

```xml
<mixture thermo_db="NASA-9" state_model="EquilTP" mechanism="none">

  <species>
    C(gr) C H O N CH CH4 CO CO2 CN C2H C2H2,acetylene C3 C3H C4 
    C4H2,butadiyne C5 HCN H2 H2O N2
  </species>

  <element_compositions default="VKIZuramPyroGas">
    <composition name="VKIZuramPyroGas">
      C:0.125, H:0.750, O:0.125, N:0.0
    </composition>
    <composition name="BLedge">
      C:0., H:0., O:0.21, N:0.79
    </composition>
    <composition name="Char">
      C:1., H:0., O:0., N:0.0
    </composition>
  </element_compositions>
</mixture>
```

**Comment by turchia** [comment]
Hello @jbscoggi,
We are here with @grgbellasvki working on this bug.
We did what suggested and few other minor adjustments and the code seems to run.
However, if we compute the viscosity including in the mixture a condensed species we get a different result than if we do not include any condensed species.
We would assume that there's a problem with multiple quantities like mass and mole fractions, pressure and number density, that are all computed for the whole mixture (from thermo) including the condensed phase. We recommend to renormalize everything wrt the gas phase only.
If you agree, we would proceed later this week.
Let us know if you think we're right. If we're not, it's @grgbellasvki fault.

**Comment by jbscoggi** [comment]
Hey @turchia and @grgbellasvki.  An easy check to confirm your hypothesis is to change 

```c++
mfd = GLOBAL
```

to 

```c++
mfd = IN_PHASE
```

in the following location:

https://github.com/mutationpp/Mutationpp/blob/d15a907ea3478694bd4ce8899d863185fb95b0e6/src/thermo/MultiPhaseEquilSolver.h#L147-L160

This will ensure mole (and mass) fractions are defined in phase, and not globally, meaning that they will sum to 1 in each phase.  Recompile and then check if you then get the same values for properties with and without the condensed phase present.  If so, then you are right.  If not, then something else is going on...

**Comment by turchia** [comment]
Hi @jbscoggi,

As @grgbellasvki pointed out this morning, the check I was doing was nonsense!  Of course, running w/ or w/o the condensed phase changed also the gas composition.
A better check (see below) confirms that indeed everything is working properly. So the fix with nGas you proposed earlier seems to work. 
We will clean up everything  and commit it.

```
turchi@arlin35 Test]$ mppequil -T 300 -P 101325 -s 0  --elem-x  "C:0.750, H:0.125, O:0.125" --scientific 3 H-O-C.xml 
      "X_C"      "X_H"      "X_O"     "X_H2"     "X_O2"  "X_C(gr)"
 2.146e-118  4.242e-37  4.733e-42  7.143e-02  7.143e-02  8.571e-01
[turchi@arlin35 Test]$ mppequil -T 300 -P 101325 -m 32  --elem-x  "C:0.750, H:0.125, O:0.125" --scientific 3 H-O-C.xml 
   mu[Pa-s]
  1.855e-05
[turchi@arlin35 Test]$ mppequil -T 300 -P 101325 -s 0  --elem-x  "H:0.125, O:0.125" --scientific 3 H-O.xml 
      "X_H"      "X_O"     "X_H2"     "X_O2"
  2.969e-36  3.313e-41  5.000e-01  5.000e-01
[turchi@arlin35 Test]$ mppequil -T 300 -P 101325 -m 32  --elem-x  "H:0.125, O:0.125" --scientific 3 H-O.xml 
   mu[Pa-s]
  1.855e-05
[turchi@arlin35 Test]$ 
```

**Comment by jbscoggi** [comment]
Hey @turchia and @grgbellasvki, just checking on the progress of this issue.  Is it OK to close?

**Comment by jbscoggi** [comment]
Fixed in #43. Closing.

**Comment by turchia** [comment]
I let  @grgbellasvki <https://github.com/grgbellasvki> give our final word...
